scalding-core/src/main/scala/com/twitter/scalding/serialization/Serialization.scala

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * This is a base Input/OutputStream-based serialization typeclass
+ * This is useful for value serialization in hadoop when we don't
+ * need to do key sorting for partitioning.
+ *
+ * This serialization typeclass must serialize equivalent objects
+ * identically to be lawful. Given that constraint, we can always
+ * get an Equiv and Hashing from a Serialization (by doing byte-wise
+ * equivalence or byte-wise hashing).
+ *
+ * A serialization always gives a hash because one can just
+ * serialize and then hash the bytes. You might prefer another
+ * implementation. This must satisfy:
+ *   (!equiv(a, b)) || (hash(a) == hash(b))
+ */
+trait Serialization[T] extends Equiv[T] with Hashing[T] with Serializable {
+  def read(in: InputStream): Try[T]
+  def write(out: OutputStream, t: T): Try[Unit]
+  /**
+   * If all items have a static size, this returns Some, else None
+   */
+  def staticSize: Option[Int] = None
+  /**
+   * returns Some if the size is cheap to calculate.
+   * otherwise the caller should just serialize into an ByteArrayOutputStream
+   */
+  def dynamicSize(t: T): Option[Int] = None
+}
+
+object Serialization {
+  import JavaStreamEnrichments._
+  /**
+   * This is a constant for us to reuse in Serialization.write
+   */
+  val successUnit: Try[Unit] = Success(())
+
+  def equiv[T](a: T, b: T)(implicit ser: Serialization[T]): Boolean =
+    ser.equiv(a, b)
+
+  def hash[T](t: T)(implicit ser: Serialization[T]): Int =
+    ser.hash(t)
+
+  def read[T](in: InputStream)(implicit ser: Serialization[T]): Try[T] =
+    ser.read(in)
+
+  def write[T](out: OutputStream, t: T)(implicit ser: Serialization[T]): Try[Unit] =
+    ser.write(out, t)
+
+  def toBytes[T: Serialization](t: T): Array[Byte] = {
+    val baos = new ByteArrayOutputStream
+    write(baos, t).get // this should only throw on OOM
+    baos.toByteArray
+  }
+
+  def fromBytes[T: Serialization](b: Array[Byte]): Try[T] =
+    read(new ByteArrayInputStream(b))
+
+  private def roundTrip[T](t: T)(implicit ser: Serialization[T]): T = {
+    val baos = new ByteArrayOutputStream
+    ser.write(baos, t).get // should never throw on a ByteArrayOutputStream
+    ser.read(baos.toInputStream).get
+  }
+
+  /**
+   * Do these two items write equivalently?
+   */
+  def writeEquiv[T: Serialization](a: T, b: T): Boolean =
+    java.util.Arrays.equals(toBytes(a), toBytes(b))
+
+  /**
+   * write followed by read should give an equivalent T
+   *
+   * This is a law that serialization must follow. It is here for
+   * documentation and for use within tests without any dependence on
+   * specific test frameworks.
+   *
+   * forAll(roundTripLaw[T]) in a valid test in scalacheck style
+   */
+  def roundTripLaw[T: Serialization]: Law1[T] =
+    Law1("roundTrip", { (t: T) => equiv(roundTrip(t), t) })
+
+  /**
+   * If two items are equal, they should serialize byte for byte equivalently
+   */
+  def serializationIsEquivalence[T: Serialization]: Law2[T] =
```

Comment:
yeah, I'm thinking actually this may be too strict. If you are this strict, I'm not sure why you would not want to go all the way to OrderedSerialization.

Dropping this strictness might be something we would do in the future if we require Serialization for all Values.